src/vs/workbench/contrib/debug/browser/debugCommands.ts

```diff
@@ -179,6 +187,101 @@ async function changeDebugConsoleFocus(accessor: ServicesAccessor, next: boolean
 	}
 }
 
+async function navigateCallStack(debugService: IDebugService, down: boolean) {
+	const frame = debugService.getViewModel().focusedStackFrame;
+	if (frame) {
+
+		let callStack = frame.thread.getCallStack();
+		const index = callStack.findIndex(elem => elem.frameId === frame.frameId);
+		let nextVisibleFrame;
+		if (down) {
+			if (index >= callStack.length - 1) {
+				if ((<Thread>frame.thread).reachedEndOfCallStack) {
+					goToTopOfCallStack(debugService);
+					return;
+				} else {
+					await debugService.getModel().fetchCallstack((<Thread>frame.thread), 20);
+					callStack = frame.thread.getCallStack();
+				}
+			}
+			nextVisibleFrame = findNextVisibleFrame(true, callStack, index);
+		} else {
+			if (index <= 0) {
+				goToBottomOfCallStack(debugService);
+				return;
+			}
+			nextVisibleFrame = findNextVisibleFrame(false, callStack, index);
+		}
+
+		if (nextVisibleFrame) {
+			debugService.focusStackFrame(nextVisibleFrame);
+		}
+	}
+}
+
+async function goToBottomOfCallStack(debugService: IDebugService) {
+	const thread = debugService.getViewModel().focusedThread;
+	if (thread) {
+		await debugService.getModel().fetchCallstack((<Thread>thread));
+		const callStack = thread.getCallStack();
+		if (callStack.length > 0) {
+			const nextVisibleFrame = findNextVisibleFrame(false, callStack, 0); // must consider the next frame up first, which will be the last frame
+			if (nextVisibleFrame) {
+				debugService.focusStackFrame(nextVisibleFrame);
+			}
+		}
+	}
+}
+
+function goToTopOfCallStack(debugService: IDebugService) {
+	const thread = debugService.getViewModel().focusedThread;
+
+	if (thread) {
+		debugService.focusStackFrame(thread.getTopStackFrame());
+	}
+}
+
+/**
+ * Finds next frame that is not skipped by SkipFiles. Skips frame at index and starts searching at next.
+ * @param down specifies whether to search downwards if the current file is skipped.
+ * @param callStack the call stack to search
+ * @param startIndex the index to start the search at
+ */
+function findNextVisibleFrame(down: boolean, callStack: readonly IStackFrame[], startIndex: number) {
+
+	if (startIndex === callStack.length) {
```

Comment:
Seems like you only pass in something between 0 and callstack.length-1, will this block be hit?

Reply:
it won't currently be hit, but it's just there to make sure that any other calls won't trigger an infinite loop. alternatively, I can just put a precondition for the function docs

Reply:
Hmm.. now that I think about it, I should probably have a check for out-of-bounds for `0 <= startIndex <= callStack - 1` in case fetching a new callstack somehow receives a shorter callstack and the index is now out-of-range. Is it possible for the new callstack to be shorter?

Reply:
I'll just fetch the index again when I refresh the callstack I think

Reply:
If you are trying to validate the startIndex input, you should check with < and > rather than one speciic number

Reply:
If it would trigger an infinite loop and make vscode hang, it's probably worth doing the full validation

Reply:
I changed it so that the caller makes sure that the index is in range. Each time an index is created, it's directly taken from searching the copy of the callstack we currently have. I could also just re-add the check. And yeah, I accidentally just used `===` instead of `<` and `>` before, will re-add